Module 1

jianingxu1 opened the pull request:
The program fetches data from the meteo API for each city, processes the data, and plots for each city and weather variable, the evolution of the mean and std of that weather variable for every climate model, to see the differences between every model. This is an example for "Madrid" and the "temperature_2m_mean" weather variable.
![madrid-temperature](https://github.com/jianingxu1/zrive-ds/assets/58917306/2669efbd-6a80-4282-a45a-6cd548c14a72)

Comments:
- The fetching of data only works for Madrid for the first time, eventually leading to error 429 "too many requests" because of the huge amount of data that it requests and the rate limit the API has. The approach to fix it has been commented on in class but I have not implemented it.

This is my first time working on Data Science projects and also, having my code reviewed, so I would love to hear ALL kinds of feedback to improve and learn. Don't hesitate to comment!

jianingxu1 committed:
Update sklearn to scikit-learn in poetry

jianingxu1 committed:
Implement fetching of meteo data

jianingxu1 committed:
Add meteo data transformations into dataframes

jianingxu1 committed:
Add plotting of meteo data

jianingxu1 committed:
Add typing, refactor code and improve plot

sergiorozada12 commented on src/module_1/module_1_meteo_api.py:
```diff
@@ -1,7 +1,169 @@
+VARIABLES = "temperature_2m_mean,precipitation_sum,soil_moisture_0_to_10cm_mean"
+
+
+def fetch_url(url: str, params: Dict[str, str] = None) -> Optional[Dict]:
```
Well done with the typings, but you should be consistent. Sometimes you specify the types in the ```Dict```, and sometimes you don't. If you type it, then you can use mipy or similar tools more efficiently.

sergiorozada12 commented on src/module_1/module_1_meteo_api.py:
```diff
@@ -1,7 +1,169 @@
+
+    df = pd.DataFrame(daily_data)
+
+    # Convert 'time' to datetime
```
Are all these comments useful? You can infer what the code is doing from the code. If something is not adding value to your code, it can go away.

sergiorozada12 commented on src/module_1/module_1_meteo_api.py:
```diff
@@ -1,7 +1,169 @@
+            print("Data fetched successfully!")
+            return response.json()
+
+        except requests.exceptions.RequestException as error:
```
Pretty nice error handling. A good practice is to retry with backoff strategies to rate limit the calls and avoid collapsing the server side:

https://en.wikipedia.org/wiki/Exponential_backoff

sergiorozada12 commented on src/module_1/module_1_meteo_api.py:
```diff
@@ -1,7 +1,169 @@
+    Fetch weather data from meteo API given a specified location,
+    time range in "YYYY-MM-DD" format and climate models.
+    """
+    params = {
```
This is perfectly fine. If you wanna explore how to design data models in a principled way take a look at dataclasses. They are specially useful for handling and designing data models:

https://docs.python.org/3/library/dataclasses.html

sergiorozada12 commented on src/module_1/module_1_meteo_api.py:
```diff
@@ -1,7 +1,169 @@
+        city_to_data[city] = data
+
+    city_to_processed_data = {}
+    for city, data in city_to_data.items():
```
Why not merging all these for loops?. Not really relevant here, but you are iterating the cities three times. What if we have 1e6 cities?

sergiorozada12 reviewed (approved):
Pretty nice PR. There's not need to submit again the PR with changes, you can merge it. However, bear in mind the comments for the following PRs. 

I strongly recommend you to go through the testing part. It's by far the most interesting part of this lab.

jianingxu1 committed:
Clean and refactor code

- Specify types in Dict and function return values
- Delete useless comments
- Refactor city loops

jianingxu1 committed:
Change to exponential backoff when retrying requests in fetch_url

jianingxu1 committed:
Use logging module instead of print

jianingxu1 committed:
Add tests

jianingxu1 commented:
In addition to implementing some of the improvements based on your feedback, I've integrated unit testing into the program, and would really appreciate a review of it (the testing part).

sergiorozada12 commented on tests/module_1/test_meteo_api.py:
```diff
@@ -1,7 +1,116 @@
+        }
+    )
+
+    pd.testing.assert_frame_equal(
```
Clean, nice test

sergiorozada12 commented on tests/module_1/test_meteo_api.py:
```diff
@@ -1,7 +1,116 @@
+
+    response = fetch_url("mock_url")
+
+    assert response == "mocked_response"
```
Perhaps overkill in this case, but in a real test check also the status code and the reason. If they are not relevant in your test, then the ```MockResponse``` class needs default values for them.

jianingxu1 replied:
Yeah, I thought about doing that but my `fetch_url` function returns directly the `response.json()` data instead of the response itself, so I cannot access the `response.status_code`.

Does this mean that in terms of design, it would be better for the `fetch_url` function to return the `response` instead of `response.json()`, thus, it is easier to test?

sergiorozada12 replied:
Depends a lot on the context, but generally speaking yes. Statuses can tell you what flow to trigger for example, so they might be useful. 

In my case I like creating a dataclass or similar to load the info from response, with status being a field. Then, you don't answer back with a general response nor a native dict, but a custom class following the schema I want.

sergiorozada12 commented on tests/module_1/test_meteo_api.py:
```diff
@@ -1,7 +1,116 @@
+        "Error 429: mocked_reason. Data fetching unsuccessful.",
+    ]
+
+    assert [r.msg for r in caplog.records] == expected_logs
```
A couple of comments here. First, this test implies ```requests.RequestException``` has been raised, so it should be under the ```pytest.raises``` context. Then, the most natural way of checking the error is by looking at the status code, rather than the log. This being said, checking the log is nice too.

sergiorozada12 commented on tests/module_1/test_meteo_api.py:
```diff
@@ -1,7 +1,116 @@
+    ) -> None:
+        self.json_data = json_data
+        self.status_code = status_code
+        self.reason = reason
```
Are reason and headers really needed here?

jianingxu1 replied:
I had to add them because during `fetch_url`, `reason` is used when logging warnings and `headers` is accessed to get the `Retry-After` attribute, and without adding them, the tests wouldn't run for those cases that used them. Is that alright?

sergiorozada12 replied:
You are fully right

sergiorozada12 reviewed (approved):
LGTM. Left a couple of nits, but it's a really good PR

jianingxu1 committed:
Create dataclass for meteo api parameters

jianingxu1 committed:
Delete extra empty plot figure
